internal/storage/fs/snapshot.go

```diff
@@ -1,14 +1,616 @@
 package fs
 
 import (
+	"context"
+	"encoding/json"
 	"errors"
+	"fmt"
 	"io"
+	iofs "io/fs"
+	"path"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/gofrs/uuid"
+	ferrors "go.flipt.io/flipt/errors"
+	"go.flipt.io/flipt/internal/ext"
+	"go.flipt.io/flipt/internal/storage"
+	"go.flipt.io/flipt/rpc/flipt"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	indexFile = ".flipt.yml"
+	defaultNs = "default"
 )
 
-type storeSnapshot struct{}
+var ErrNotImplemented = errors.New("not implemented")
+
+// FliptIndex represents the structure of a well-known file ".flipt.yml"
+// at the root of an FS.
+type FliptIndex struct {
+	Version    string   `yaml:"version,omitempty"`
+	Inclusions []string `yaml:"inclusions,omitempty"`
+	Exclusions []string `yaml:"exclusions,omitempty"`
+}
+
+// storeSnapshot contains the structures necessary for serving
+// flag state to a client.
+type storeSnapshot struct {
+	ns        map[string]*namespace
+	evalDists map[string][]*storage.EvaluationDistribution
+}
+
+type namespace struct {
+	resource  *flipt.Namespace
+	flags     map[string]*flipt.Flag
+	segments  map[string]*flipt.Segment
+	rules     map[string]*flipt.Rule
+	evalRules map[string][]*storage.EvaluationRule
+}
+
+func newNamespace(key string) *namespace {
+	return &namespace{
+		resource:  &flipt.Namespace{Key: key, Name: strings.Title(key)},
+		flags:     map[string]*flipt.Flag{},
+		segments:  map[string]*flipt.Segment{},
+		rules:     map[string]*flipt.Rule{},
+		evalRules: map[string][]*storage.EvaluationRule{},
+	}
+}
 
 // snapshotFromFS constructs a storeSnapshot from the provided
 // fs.FS implementation.
 func snapshotFromFS(sources ...io.Reader) (*storeSnapshot, error) {
```

Comment:
I know that I contributed this function, but just while you're here, could you rename this:
```suggestion
// snapshotFromReaders constructs a storeSnapshot from the provided
// slice of io.Reader.
func snapshotFromReaders(sources ...io.Reader) (*storeSnapshot, error) {
```